### PR TITLE
[RHPAM-3640] Configure Kafka JBPM Emitter from Operator

### PR DESCRIPTION
**Pull request by desmax74**
See:https://issues.redhat.com/browse/RHPAM-3640
Signed-off-by: desmax74 <mdessi@redhat.com>
7.12.0

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [x] Pull Request title is properly formatted: `[KIECLOUD-XYZ] Subject`, `[RHDM-XYZ] Subject` or `[RHPAM-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted
- [x] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [x] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 1]
```diff
@@ -792,3 +799,40 @@ function configure_kafka(){
...
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_DATE_FORMAT}" ];then
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.date_format=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_DATE_FORMAT}"
+          fi
+
+          JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.topic.processes=jbpm-process-events"
+          JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.topic.tasks=jbpm-task-events"
+          JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.topic.cases=jbpm-case-events"
+
+          if [ -d "/opt/kie/dependencies/jbpm-kafka" ] ;then
+            for f in /opt/kie/dependencies/jbpm-kafka/jbpm-event-emitters-kafka-*.jar; do
```
Comment (spolti) [inline jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:828]: do we expect to have more than 1 emitter jar?
if not remove the for loop.

File: jboss-kie-kieserver/configure.sh [hunk 2]
```diff
@@ -61,3 +61,8 @@ chmod -R 777 /opt/kie/dependencies/jbpm-clustering
 # Create dir to remove JDBC driver
 mkdir ${JBOSS_HOME}/modules/system/layers/openshift 2&> /dev/null || true
 chown -R jboss:root ${JBOSS_HOME}/modules/system/layers/openshift
+
+# Dir for optional deps
+mkdir -p /opt/kie/dependencies/jbpm-kafka
+chown -R jboss:root /opt/kie/dependencies/jbpm-kafka
+chmod -R 777 /opt/kie/dependencies/jbpm-kafka
```
Comment (spolti) [inline jboss-kie-kieserver/configure.sh:68]: please do not use 777, it should be 755.
also, please remove it from **chmod -R 777 /opt/kie/dependencies/jbpm-clustering**
and other places in this script.

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 3]
```diff
@@ -792,3 +799,40 @@ function configure_kafka(){
...
 }
+
+function configure_kafka_jbpm_emitter(){
+  if [ "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ENABLED^^}" = "TRUE" ]; then
+      if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}" ];then
+          log_info "Kafka JBPM Emitter enabled"
+          # Replace with the follow when this https://issues.redhat.com/browse/JBPM-9716 will be merged and the jar will be realigned
+          #JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.bootstrap.servers=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}"
+          JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.boopstrap.servers=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}"
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS}" ];then
```
Comment (spolti) [inline jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:810]: default value?

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 4]
```diff
@@ -792,3 +799,40 @@ function configure_kafka(){
...
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.acks=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS}"
+          fi
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CLIENT_ID}" ];then
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.client.id=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CLIENT_ID}"
+          fi
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_MAX_BLOCK_MS}" ];then
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.max.block.ms=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_MAX_BLOCK_MS}"
+          fi
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_DATE_FORMAT}" ];then
```
Comment (spolti) [inline jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:819]: same here.

File: tests/features/common/kie-kieserver-common.feature [hunk 5]
```diff
@@ -570,3 +570,37 @@ Feature: Kie Server common features
...
+      | variable                                              | value                         |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS | localhost:9093                |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ENABLED           | true                          |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CLIENT_ID         | jbpmapp                       |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS              | 3                             |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_MAX_BLOCK_MS      | 2100                          |
+      | KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_DATE_FORMAT       | dd-MM-yyyy'T'HH:mm:ss.SSSZ    |
+      | SCRIPT_DEBUG                                          | true                          |
+    Then container log should contain -Dorg.kie.kafka.server.ext.disabled=true
+    # Replace with the follow when this https://issues.redhat.com/browse/JBPM-9716 will be merged and the jar will be realigned
```
Comment (spolti) [inline tests/features/common/kie-kieserver-common.feature:585]: jira is closed, is there anything else missing?

**Review by spolti — CHANGES_REQUESTED** [review]

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 6]
```diff
@@ -36,6 +36,12 @@ function prepareEnv() {
     unset KIE_SERVER_KAFKA_EXT_GROUP_ID
     unset KIE_SERVER_KAFKA_EXT_MAX_BLOCK_MS
     unset KIE_SERVER_KAFKA_EXT_TOPICS
+    unset KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS
```
Comment (spolti) [inline jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:39]: ANd, please, document this envs on the image.yaml descriptor for the kieserver image.

**Review by spolti — CHANGES_REQUESTED** [review]

File: jboss-kie-kieserver/configure.sh (hunk 2 above)
Comment (desmax74) [reply jboss-kie-kieserver/configure.sh:68]: ok

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 1 above)
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:828]: Missing align with 7.11.1 PR

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 7]
```diff
@@ -792,3 +799,40 @@ function configure_kafka(){
     JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.kafka.server.ext.disabled=true"
   fi
 }
+
+function configure_kafka_jbpm_emitter(){
+  if [ "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ENABLED^^}" = "TRUE" ]; then
+      if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}" ];then
+          log_info "Kafka JBPM Emitter enabled"
+          # Replace with the follow when this https://issues.redhat.com/browse/JBPM-9716 will be merged and the jar will be realigned
```
[the remark this answers is not in the document]
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh]: Aligned with 7.11.1

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 3 above)
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:808]: On acks no default values in the emitter's code

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 4 above)
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:817]: Default value if missing is picked from org.kie.server.json.date_format in the emitter's code

File: tests/features/common/kie-kieserver-common.feature (hunk 5 above)
Comment (desmax74) [reply tests/features/common/kie-kieserver-common.feature:585]: No

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh [hunk 8]
```diff
@@ -792,3 +799,40 @@ function configure_kafka(){
...
+          #JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.bootstrap.servers=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}"
+          JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.boopstrap.servers=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_BOOTSTRAP_SERVERS}"
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS}" ];then
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.acks=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_ACKS}"
+          fi
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CLIENT_ID}" ];then
+            JBOSS_KIE_ARGS="${JBOSS_KIE_ARGS} -Dorg.kie.jbpm.event.emitters.kafka.client.id=${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CLIENT_ID}"
+          fi
+          if [ -n "${KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_MAX_BLOCK_MS}" ];then
```
[the remark this answers is not in the document]
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:814]: ok

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 6 above)
Comment (desmax74) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:39]: https://github.com/jboss-container-images/rhpam-7-openshift-image/pull/551

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 3 above)
Comment (spolti) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:808]: why not add here as well to save a few lines of code?

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 4 above)
Comment (spolti) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:817]: is there an y validation to make sure the value provided is valid?

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 3 above)
Comment (spolti) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:808]: ok, make sure to add what is the default value on the image.yaml.

File: jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh (hunk 4 above)
Comment (spolti) [reply jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh:817]: plus, make sure to add what is the default value on the image.yaml.

**Review by spolti — APPROVED** [review]

**Comment by spolti** [comment]
hold for now, will merge all PRs tomorrow morning.

File: jboss-kie-kieserver/tests/bats/jboss-kie-kieserver.bats [hunk 9]
```diff
@@ -715,3 +715,64 @@ EOF
...
+    configure_kafka_jbpm_emitter
+
+    local expected=" -Dorg.kie.jbpm.event.emitters.kafka.bootstrap.servers=localhost:9093 -Dorg.kie.jbpm.event.emitters.kafka.max.block.ms=2100 -Dorg.kie.jbpm.event.emitters.kafka.acks=3 -Dorg.kie.jbpm.event.emitters.kafka.client.id=jbpmapp -Dorg.kie.jbpm.event.emitters.kafka.date_format=dd-MM-yyyy'T'HH:mm:ss.SSSZ -Dorg.kie.jbpm.event.emitters.kafka.topic.processes=my-processes-topic -Dorg.kie.jbpm.event.emitters.kafka.topic.tasks=my-tasks-topic -Dorg.kie.jbpm.event.emitters.kafka.topic.cases=my-cases-topic"
+    echo "  Result: ${JBOSS_KIE_ARGS}"
+
+    echo "Expected: ${expected}"
+    [[ "${JBOSS_KIE_ARGS}" == "${expected}" ]]
+}
+
+@test "Verify the Kafka JBPM Emitter without bootstrap emitter" {
```
Comment (spolti) [inline jboss-kie-kieserver/tests/bats/jboss-kie-kieserver.bats:739]: whats the purpose of this test?

**Review by spolti — APPROVED** [review]

**Pull request merged**